### PR TITLE
 Flexible perception pipeline manipulation for RoboSherlock An T. Le GSOC 2018

### PR DESCRIPTION
**Pull request by anindex**
### Summary
**Features**
- Implement annotator dependencies query interface from knowrob to plan parallel pipeline
- Implement ordering-based parallel pipeline planner using annotator dependencies
- Implement parallel pipeline execution capability

**Modules implemented**
- DirectedGraph: add a variant of graph into robosherlock's graph lib, data structure needed for dependency planning 
- RSParallelPipelinePlanner: query annotator dependencies from knowrob and plan dependency orderings
- RSParallelAnalysisEngine: port to RSControledAnalysisEngine and execute parallel pipeline using dependency orderings. It is inherited from uima::AggregatedEngine for backward compatible with linear execution.

**Usage**
Setup steps:
1) Run bag files or connect to kinect
2) Setup knowrob: `roslaunch json_prolog json_prolog.launch initial_package:=robosherlock_knowrob
`
3) Run demo pipeline: `roslaunch robosherlock rs.launch ae:=demo wait:=false`

For linear execution, run `roslaunch robosherlock rs.launch ae:=demo wait:=false parallel:=false`

### CODE REVIEW

**Comment by bbferka** [comment]
@anindex check the build log on travis pls. your PR is incomplete.

**Comment by anindex** [comment]
Oh yes, because I have changed the flowcontrol files. Therefore, the unit tests are affected. I will fix them tomorrow :)

**Comment by anindex** [comment]
Weird.. catkin tools build and test success but catkin_make run_tests in travis failed

**Comment by bbferka** [comment]
@anindex hm, can you think of tests where you don't need Prolog? e.g. some way of replicating these results?

**Comment by anindex** [comment]
So in order to merge this PR we have to turn off the codecov right?

File: src/flowcontrol/include/rs/flowcontrol/RSPipelineManager.h [hunk 1]
```diff
@@ -63,10 +84,13 @@ class RSPipelineManager
...
   int getIndexOfAnnotator(std::string annotator_name);
 
+  void getCurrentAnnotatorFlow(std::vector<std::string> &annotators);
+
+public:
   // private:
   /* data */
   uima::AnalysisEngine *engine;
-  uima::internal::AggregateEngine *aengine;
+  RSParallelAnalysisEngine *aengine;
```
Comment (bbferka) [inline src/flowcontrol/include/rs/flowcontrol/RSPipelineManager.h]: why do we have two engines here?; Couldn't we do this with a single one?

File: src/queryanswering/src/QueryInterface.cpp [hunk 2]
```diff
@@ -293,4 +294,20 @@ void QueryInterface::filterResults(std::vector<std::string> &resultDesignators,
...
+      success = jsonPrologInterface->retrieveAnnotatorsInputOutput(annotators, dependencies);
+    }
+    catch(...)
+    {
+      outError("jsonPrologInterface query failed!");
+    }
+
+    return success;
+}
+
```
Comment (bbferka) [inline src/queryanswering/src/QueryInterface.cpp]: consider keeping this in the JsonPrologInterface class;the QueryInterface only serves for handling the input queries to RoboSherlock;

File: src/flowcontrol/include/rs/flowcontrol/RSParallelAnalysisEngine.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+#include <memory>
+#include <vector>
+#include <utility>
+#include <future>
+#include <functional>
+#include <mutex>
+#include <assert.h>
+
+class RSParallelAnalysisEngine : public uima::internal::AggregateEngine
```
Comment (bbferka) [inline src/flowcontrol/include/rs/flowcontrol/RSParallelAnalysisEngine.h]: let's rename this to RSAggregateAnalysisEngine, and see if uima::AnalysisEngine could be replaced by this guy everywhere;

File: src/flowcontrol/include/rs/flowcontrol/RSPipelineManager.h (hunk 1 above)
Comment (anindex) [reply src/flowcontrol/include/rs/flowcontrol/RSPipelineManager.h]: It's just the pointer of child class and parent class pointed to the same object. I respect the old implementation of RSPipelineManager, it is from the original file. The reason is that you dont have to cast up or down everytime needed

File: src/queryanswering/src/QueryInterface.cpp (hunk 2 above)
Comment (anindex) [reply src/queryanswering/src/QueryInterface.cpp]: ok, i wil change it tomorrow

File: src/flowcontrol/include/rs/flowcontrol/RSParallelAnalysisEngine.h (hunk 3 above)
Comment (anindex) [reply src/flowcontrol/include/rs/flowcontrol/RSParallelAnalysisEngine.h]: ok, I will do it

File: descriptors/analysis_engines/demo.xml [hunk 4]
```diff
@@ -151,7 +151,7 @@
         <name>camera_config_files</name>
         <value>
           <array>
-            <string>config_kinect_robot.ini</string>
+            <string>config_data_loader.ini</string>
```
Comment (bbferka) [inline descriptors/analysis_engines/demo.xml]: you have to change this back, breaks the tutorials

**Pull request merged**